README.md
**ARCHIVED**: This is no longer maintained to focus on other packages.

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**

- [automated_testing_framework_plugin_github](#automated_testing_framework_plugin_github)
  - [Table of Contents](#table-of-contents)
  - [Introduction](#introduction)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

# automated_testing_framework_plugin_github

## Table of Contents

* [Introduction](#introduction)


## Introduction

A test store with the ability to read and write tests and images using GitHub.  See the example app for usage info.



CHANGELOG.md
## [2.0.1+15] - May 9, 2023

* Automated dependency updates


## [2.0.1+14] - May 2, 2023

* Automated dependency updates


## [2.0.1+13] - April 25, 2023

* Automated dependency updates


## [2.0.1+12] - April 18, 2023

* Automated dependency updates


## [2.0.1+11] - April 11, 2023

* Automated dependency updates


## [2.0.1+10] - April 4, 2023

* Automated dependency updates


## [2.0.1+9] - March 28, 2023

* Automated dependency updates


## [2.0.1+8] - March 21, 2023

* Automated dependency updates


## [2.0.1+7] - March 14, 2023

* Automated dependency updates


## [2.0.1+6] - March 7, 2023

* Automated dependency updates


## [2.0.1+5] - February 28, 2023

* Automated dependency updates


## [2.0.1+4] - February 21, 2023

* Automated dependency updates


## [2.0.1+3] - February 14, 2023

* Automated dependency updates


## [2.0.1+2] - February 7, 2023

* Automated dependency updates


## [2.0.1+1] - January 31, 2023

* Automated dependency updates


## [2.0.1] - January 24th, 2023

* Flutter 3.7


## [2.0.0+29] - January 24, 2023

* Automated dependency updates


## [2.0.0+28] - January 17, 2023

* Automated dependency updates


## [2.0.0+27] - January 10, 2023

* Automated dependency updates


## [2.0.0+26] - January 3, 2023

* Automated dependency updates


## [2.0.0+25] - December 27, 2022

* Automated dependency updates


## [2.0.0+24] - December 21, 2022

* Automated dependency updates


## [2.0.0+23] - December 13, 2022

* Automated dependency updates


## [2.0.0+22] - December 6, 2022

* Automated dependency updates


## [2.0.0+21] - November 22, 2022

* Automated dependency updates


## [2.0.0+20] - November 15, 2022

* Automated dependency updates


## [2.0.0+19] - November 8, 2022

* Automated dependency updates


## [2.0.0+18] - November 1, 2022

* Automated dependency updates


## [2.0.0+17] - October 25, 2022

* Automated dependency updates


## [2.0.0+16] - October 18, 2022

* Automated dependency updates


## [2.0.0+15] - September 27, 2022

* Automated dependency updates


## [2.0.0+14] - September 20, 2022

* Automated dependency updates


## [2.0.0+13] - September 13, 2022

* Automated dependency updates


## [2.0.0+12] - September 6, 2022

* Automated dependency updates


## [2.0.0+11] - August 30, 2022

* Automated dependency updates


## [2.0.0+10] - August 16, 2022

* Automated dependency updates


## [2.0.0+9] - August 2, 2022

* Automated dependency updates


## [2.0.0+8] - July 26, 2022

* Automated dependency updates


## [2.0.0+7] - July 19, 2022

* Automated dependency updates


## [2.0.0+6] - June 28, 2022

* Automated dependency updates


## [2.0.0+5] - June 21, 2022

* Automated dependency updates


## [2.0.0+4] - June, 14, 2022

* Automated dependency updates


## [2.0.0+3] - June, 7, 2022

* Automated dependency updates


## [2.0.0+2] - May, 30, 2022

* Automated dependency updates


## [2.0.0+1] - May 14th, 2022

* Flutter 3.0


## [1.0.0+3] - April 16th, 2022

* Dependency updates


## [1.0.0+2] - February 16th, 2022

* Real release


## [0.9.0] - January 20th, 2022

* Beta release













































example/lib/main.dart
import 'dart:io';

import 'package:automated_testing_framework/automated_testing_framework.dart';
import 'package:automated_testing_framework_example/automated_testing_framework_example.dart';
import 'package:automated_testing_framework_plugin_github/automated_testing_framework_plugin_github.dart';
import 'package:automated_testing_framework_plugin_images/automated_testing_framework_plugin_images.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:github/github.dart';
import 'package:logging/logging.dart';

void main() async {
  Logger.root.level = Level.ALL;
  Logger.root.onRecord.listen((record) {
    // ignore: avoid_print
    print('${record.level.name}: ${record.time}: ${record.message}');
    if (record.error != null) {
      // ignore: avoid_print
      print('${record.error}');
    }
    if (record.stackTrace != null) {
      // ignore: avoid_print
      print('${record.stackTrace}');
    }
  });

  WidgetsFlutterBinding.ensureInitialized();

  TestImagesHelper.registerTestSteps();

  GitHub? github;
  String? token;
  try {
    token = (await rootBundle.loadString('assets/secrets/token.txt')).trim();

    github = GitHub(auth: Authentication.withToken(token));
  } catch (e) {
    // no-op; use default auth
  }
  github ??= GitHub();

  var gestures = TestableGestures();
  if (kIsWeb ||
      Platform.isFuchsia ||
      Platform.isLinux ||
      Platform.isMacOS ||
      Platform.isWindows) {
    gestures = TestableGestures(
      widgetLongPress: null,
      widgetSecondaryLongPress: TestableGestureAction.open_test_actions_page,
      widgetSecondaryTap: TestableGestureAction.open_test_actions_dialog,
    );
  }

  final store = GithubTestStore(
    branch: 'tests',
    github: github,
    slug: RepositorySlug(
      'peiffer-innovations',
      'automated_testing_framework_plugin_github',
    ),
  );

  runApp(App(
    options: TestExampleOptions(
      autorun: kProfileMode,
      enabled: true,
      gestures: gestures,
      goldenImageWriter: store.goldenImageWriter,
      testImageReader: store.testImageReader,
      testReader: store.testReader,
      testWidgetsEnabled: true,
      testWriter: store.testWriter,
    ),
  ));
}


web/flutter_service_worker.js
'use strict';
const MANIFEST = 'flutter-app-manifest';
const TEMP = 'flutter-temp-cache';
const CACHE_NAME = 'flutter-app-cache';
const RESOURCES = {
  "version.json": "4b6db237b3514a88107a422469adfb0f",
"assets/FontManifest.json": "3a1c01ebd9adb7aecf384a895a3cbb43",
"assets/fonts/MaterialIcons-Regular.otf": "132a5e63b5e510933ab4845577716106",
"assets/NOTICES": "1a2e5011e94000d2e56dec15feb3243f",
"assets/AssetManifest.json": "68c85d7714e3b6e58070b7d2aa8853e7",
"assets/assets-src/MaterialIcons-Regular.ttf": "a37b0c01c0baf1888ca812cc0508f6e2",
"assets/packages/automated_testing_framework_example/assets/all_tests.json": "2d2c55430cb923f834f1683ae76deaf4",
"assets/packages/automated_testing_framework_example/assets/tests/exit_app.json": "358f6cab447478d73eea4127678106ac",
"assets/packages/automated_testing_framework_example/assets/tests/theme.json": "5b72a996ff15d64d58b7627f29df05bb",
"assets/packages/automated_testing_framework_example/assets/tests/double_tap.json": "35a920944da4cf2e2154bf1c138c2e34",
"assets/packages/automated_testing_framework_example/assets/tests/variables.json": "7d554f244b740f375961fb0c9f80b393",
"assets/packages/automated_testing_framework_example/assets/tests/slidables.json": "0e6ecffce2e4fe788c9f7a00e1a7d4e7",
"assets/packages/automated_testing_framework_example/assets/tests/accessibility.json": "c98acf550ea22677d4adfa3016b120f8",
"assets/packages/automated_testing_framework_example/assets/tests/buttons.json": "d35cccfbd9c0c4547e34441f6038ac5c",
"assets/packages/automated_testing_framework_example/assets/tests/issue_5.json": "21c374a01729bf635c57bc9effc9d37b",
"assets/packages/automated_testing_framework_example/assets/tests/icons_gesture.json": "655cf20202457e9a9b8692f05334fbc8",
"assets/packages/automated_testing_framework_example/assets/tests/failure.json": "8cefd140b688b564169ed2d304a52ce8",
"assets/packages/automated_testing_framework_example/assets/tests/dropdowns.json": "f1f2f78f6bbb729048728e05abca6d41",
"assets/packages/automated_testing_framework_example/assets/tests/interpolated_variables.json": "2491e0da34265345c86a6fe89a089894",
"assets/packages/automated_testing_framework_example/assets/tests/stacked_scrolling.json": "a98582489811cc6fc6a56cfcd5835699",
"assets/packages/automated_testing_framework_example/assets/tests/screenshot.json": "5a070f35d91e2c1cde99395809a1d876",
"main.dart.js": "afb54c4fbef8cf44f924a7990272aec6",
"flutter.js": "a85fcf6324d3c4d3ae3be1ae4931e9c5",
"canvaskit/profiling/canvaskit.js": "c21852696bc1cc82e8894d851c01921a",
"canvaskit/profiling/canvaskit.wasm": "371bc4e204443b0d5e774d64a046eb99",
"canvaskit/canvaskit.js": "97937cb4c2c2073c968525a3e08c86a3",
"canvaskit/canvaskit.wasm": "3de12d898ec208a5f31362cc00f09b9e",
"manifest.json": "15f73b7e8a8209c2206210b3ac8dea1b",
"index.html": "979598ad7970bd0881cbfc010eb9984f",
"/": "979598ad7970bd0881cbfc010eb9984f",
"favicon.png": "7d2009652f165b0bc9d9e23452621c35",
"icons/Icon-192.png": "c8302094bb862f810767070975583892",
"icons/Icon-512.png": "fcc99bfdc6a50774538d60bb93a26b0f"
};

// The application shell files that are downloaded before a service worker can
// start.
const CORE = [
  "main.dart.js",
"index.html",
"assets/AssetManifest.json",
"assets/FontManifest.json"];
// During install, the TEMP cache is populated with the application shell files.
self.addEventListener("install", (event) => {
  self.skipWaiting();
  return event.waitUntil(
    caches.open(TEMP).then((cache) => {
      return cache.addAll(
        CORE.map((value) => new Request(value, {'cache': 'reload'})));
    })
  );
});

// During activate, the cache is populated with the temp files downloaded in
// install. If this service worker is upgrading from one with a saved
// MANIFEST, then use this to retain unchanged resource files.
self.addEventListener("activate", function(event) {
  return event.waitUntil(async function() {
    try {
      var contentCache = await caches.open(CACHE_NAME);
      var tempCache = await caches.open(TEMP);
      var manifestCache = await caches.open(MANIFEST);
      var manifest = await manifestCache.match('manifest');
      // When there is no prior manifest, clear the entire cache.
      if (!manifest) {
        await caches.delete(CACHE_NAME);
        contentCache = await caches.open(CACHE_NAME);
        for (var request of await tempCache.keys()) {
          var response = await tempCache.match(request);
          await contentCache.put(request, response);
        }
        await caches.delete(TEMP);
        // Save the manifest to make future upgrades efficient.
        await manifestCache.put('manifest', new Response(JSON.stringify(RESOURCES)));
        return;
      }
      var oldManifest = await manifest.json();
      var origin = self.location.origin;
      for (var request of await contentCache.keys()) {
        var key = request.url.substring(origin.length + 1);
        if (key == "") {
          key = "/";
        }
        // If a resource from the old manifest is not in the new cache, or if
        // the MD5 sum has changed, delete it. Otherwise the resource is left
        // in the cache and can be reused by the new service worker.
        if (!RESOURCES[key] || RESOURCES[key] != oldManifest[key]) {
          await contentCache.delete(request);
        }
      }
      // Populate the cache with the app shell TEMP files, potentially overwriting
      // cache files preserved above.
      for (var request of await tempCache.keys()) {
        var response = await tempCache.match(request);
        await contentCache.put(request, response);
      }
      await caches.delete(TEMP);
      // Save the manifest to make future upgrades efficient.
      await manifestCache.put('manifest', new Response(JSON.stringify(RESOURCES)));
      return;
    } catch (err) {
      // On an unhandled exception the state of the cache cannot be guaranteed.
      console.error('Failed to upgrade service worker: ' + err);
      await caches.delete(CACHE_NAME);
      await caches.delete(TEMP);
      await caches.delete(MANIFEST);
    }
  }());
});

// The fetch handler redirects requests for RESOURCE files to the service
// worker cache.
self.addEventListener("fetch", (event) => {
  if (event.request.method !== 'GET') {
    return;
  }
  var origin = self.location.origin;
  var key = event.request.url.substring(origin.length + 1);
  // Redirect URLs to the index.html
  if (key.indexOf('?v=') != -1) {
    key = key.split('?v=')[0];
  }
  if (event.request.url == origin || event.request.url.startsWith(origin + '/#') || key == '') {
    key = '/';
  }
  // If the URL is not the RESOURCE list then return to signal that the
  // browser should take over.
  if (!RESOURCES[key]) {
    return;
  }
  // If the URL is the index.html, perform an online-first request.
  if (key == '/') {
    return onlineFirst(event);
  }
  event.respondWith(caches.open(CACHE_NAME)
    .then((cache) =>  {
      return cache.match(event.request).then((response) => {
        // Either respond with the cached resource, or perform a fetch and
        // lazily populate the cache only if the resource was successfully fetched.
        return response || fetch(event.request).then((response) => {
          if (response && Boolean(response.ok)) {
            cache.put(event.request, response.clone());
          }
          return response;
        });
      })
    })
  );
});

self.addEventListener('message', (event) => {
  // SkipWaiting can be used to immediately activate a waiting service worker.
  // This will also require a page refresh triggered by the main worker.
  if (event.data === 'skipWaiting') {
    self.skipWaiting();
    return;
  }
  if (event.data === 'downloadOffline') {
    downloadOffline();
    return;
  }
});

// Download offline will check the RESOURCES for all files not in the cache
// and populate them.
async function downloadOffline() {
  var resources = [];
  var contentCache = await caches.open(CACHE_NAME);
  var currentContent = {};
  for (var request of await contentCache.keys()) {
    var key = request.url.substring(origin.length + 1);
    if (key == "") {
      key = "/";
    }
    currentContent[key] = true;
  }
  for (var resourceKey of Object.keys(RESOURCES)) {
    if (!currentContent[resourceKey]) {
      resources.push(resourceKey);
    }
  }
  return contentCache.addAll(resources);
}

// Attempt to download the resource online before falling back to
// the offline cache.
function onlineFirst(event) {
  return event.respondWith(
    fetch(event.request).then((response) => {
      return caches.open(CACHE_NAME).then((cache) => {
        cache.put(event.request, response.clone());
        return response;
      });
    }).catch((error) => {
      return caches.open(CACHE_NAME).then((cache) => {
        return cache.match(event.request).then((response) => {
          if (response != null) {
            return response;
          }
          throw error;
        });
      });
    })
  );
}


lib/src/github_test_store.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:automated_testing_framework_models/automated_testing_framework_models.dart';
import 'package:convert/convert.dart';
import 'package:flutter/material.dart';
import 'package:github/github.dart';
import 'package:logging/logging.dart';
import 'package:pointycastle/pointycastle.dart';

class GithubTestStore {
  GithubTestStore({
    this.branch = 'main',
    String committerEmail = 'noop@github.com',
    String committerName = 'ATF GH Plugin',
    required GitHub github,
    this.goldenImagesPath = 'goldens',
    required this.slug,
    this.testsPath = 'tests',
  })  :
        // Double encode to turn it fully into an escaped JSON encoded string
        _committer = {
          'email': committerEmail,
          'name': committerName,
        },
        _github = github;

  static final Logger _logger = Logger('GithubTestStore');

  final String branch;

  final String goldenImagesPath;

  final RepositorySlug slug;

  /// Optional path to store tests.
  final String testsPath;

  final Map<String, dynamic> _committer;
  final GitHub _github;

  /// Cached value that will be refreshed as needed.
  GoldenTestImages? _currentGoldenTestImages;
  bool _dirty = true;
  late GitTree _tree;

  static String _createGoldenImageId({
    required TestDeviceInfo deviceInfo,
    String? suiteName,
    required String testName,
  }) {
    final suitePrefix = suiteName?.isNotEmpty == true
        ? '${suiteName!.replaceAll('/', '-')}/'
        : '';
    return '${deviceInfo.appIdentifier}/${suitePrefix}${testName.replaceAll('/', '-')}/${deviceInfo.os}/${deviceInfo.systemVersion}/${deviceInfo.model}/${deviceInfo.device}/${deviceInfo.orientation}/${deviceInfo.pixels?.height}x${deviceInfo.pixels?.width}';
  }

  static String _createGoldenImageIdFromReport(TestReport report) {
    final suiteName = report.suiteName;
    final testName = report.name;
    final deviceInfo = report.deviceInfo;

    return _createGoldenImageId(
      deviceInfo: deviceInfo ?? TestDeviceInfo.unknown(),
      suiteName: suiteName,
      testName: testName ?? 'unknown',
    );
  }

  /// Writes the golden images from the [report] to Cloud Storage and also
  /// writes the metadata that allows the reading of the golden images.  This
  /// will throw an exception on failure.
  Future<void> goldenImageWriter(TestReport report) async {
    await _loadData();
    final actualPath = goldenImagesPath.startsWith('/')
        ? goldenImagesPath.substring(1)
        : goldenImagesPath;

    final goldenId = _createGoldenImageIdFromReport(report);

    final data = <String, String>{};
    for (var image in report.images) {
      if (image.goldenCompatible == true) {
        data[image.id] = image.hash;
      }
    }
    final golden = GoldenTestImages(
      deviceInfo: report.deviceInfo!,
      goldenHashes: data,
      suiteName: report.suiteName,
      testName: report.name!,
      testVersion: report.version,
    );

    const batchSize = 10;
    final batch = <Future>[];

    final hashes = <String>{};

    final images = List.from(report.images);

    images.removeWhere((image) {
      var found = false;

      for (var e in _tree.entries!) {
        if (e.path == '$actualPath/$goldenId/${image.hash}.png') {
          found = true;
          break;
        }
      }

      return !found;
    });

    for (var image in report.images) {
      if (image.goldenCompatible) {
        if (!hashes.contains(image.hash)) {
          hashes.add(image.hash);
          final path = '$actualPath/$goldenId/${image.hash}.png';
          final entry = _tree.entries!.where((e) => e.path == path);

          if (entry.isNotEmpty) {
            final first = entry.first;
            final hash1 = hex.encode(Digest('SHA-1').process(image.image!));

            if (first.sha == hash1) {
              // Contents are the same, don't bother updating it.
              continue;
            }
          }

          batch.add(uploadFile(
            data: image.image!,
            message: 'Golden Image Update',
            path: path,
          ));

          if (batch.length >= batchSize) {
            await Future.wait(batch);
            batch.clear();
          }
        }
      }
    }

    batch.add(uploadFile(
      data: utf8.encode(golden.toString()),
      message: 'Golden Image Update',
      path: '$actualPath/$goldenId.json',
    ));

    await Future.wait(batch);
    batch.clear();

    _dirty = true;
  }

  /// Reader to read a golden image from GitHub.
  Future<Uint8List?> testImageReader({
    required TestDeviceInfo deviceInfo,
    required String imageId,
    String? suiteName,
    required String testName,
    int? testVersion,
  }) async {
    await _loadData();
    final goldenId = GoldenTestImages.createId(
      deviceInfo: deviceInfo,
      suiteName: suiteName,
      testName: testName,
    );
    final actualPath = goldenImagesPath.startsWith('/')
        ? goldenImagesPath.substring(1)
        : goldenImagesPath;
    GoldenTestImages? golden;
    if (_currentGoldenTestImages?.id == goldenId) {
      golden = _currentGoldenTestImages;
    } else {
      final name = '$actualPath/$goldenId.json';
      final entry = _tree.entries!.where((e) => e.path == name);

      if (entry.isNotEmpty) {
        final response = await _github.repositories.getContents(
          slug,
          _encodeGitPath(actualPath),
        );

        if (response.file?.content != null) {
          final data = utf8.decode(base64.decode(response.file!.content!));
          final goldenJson = json.decode(data);
          golden = GoldenTestImages.fromDynamic(goldenJson);
          _currentGoldenTestImages = golden;
        }
      }
    }

    Uint8List? image;
    if (golden != null) {
      final hash = golden.goldenHashes![imageId];
      final path = '$actualPath/$goldenId/$hash.png';
      final entry = _tree.entries!.where((e) => e.path == path);
      if (entry.isNotEmpty) {
        final response = await _github.repositories.getContents(
          slug,
          _encodeGitPath(path),
        );

        if (response.file?.content != null) {
          image = base64.decode(response.file!.content!);
        }
      }
    }

    return image;
  }

  /// Implementation of the [TestReader] functional interface that can read test
  /// data from GitHub.
  Future<List<PendingTest>> testReader(
    BuildContext? context, {
    String? suiteName,
  }) async {
    await _loadData();
    final results = <PendingTest>[];

    try {
      final actualTestsPath =
          testsPath.startsWith('/') ? testsPath.substring(1) : testsPath;

      final tests = _tree.entries!.where(
        (entry) =>
            (actualTestsPath.isEmpty ||
                entry.path?.startsWith(actualTestsPath) == true) &&
            entry.path!.endsWith('.json'),
      );

      final batchSize = 10;
      final batch = <Future>[];

      for (var test in tests) {
        batch.add(
          Future.microtask(() async {
            final encodedPath = _encodeGitPath(test.path!);
            try {
              final contents = await _github.repositories.getContents(
                slug,
                encodedPath,
                ref: branch,
              );

              final body = utf8.decode(
                base64.decode(
                  contents.file!.content!.replaceAll('\n', ''),
                ),
              );
              final fullTest = Test.fromDynamic(json.decode(body));
              results.add(
                PendingTest.memory(fullTest),
              );
            } catch (e, stack) {
              _logger.warning(
                '[TEST LOAD FAILED]: Failed attempting to load test at path: [${test.path}]',
                e,
                stack,
              );
            }
          }),
        );
        if (batch.length >= batchSize) {
          await Future.wait(batch);
          batch.clear();
        }
      }

      if (batch.isNotEmpty) {
        await Future.wait(batch);
        batch.clear();
      }

      results.sort(
        (a, b) => a.name.toLowerCase().compareTo(b.name.toLowerCase()),
      );
    } catch (e, stack) {
      _logger.severe('Error loading tests', e, stack);
    }

    return results;
  }

  /// Implementation of the [TestWriter] functional interface that can submit
  /// test data to GitHub.
  Future<bool> testWriter(
    BuildContext context,
    Test test,
  ) async {
    await _loadData();
    var result = false;

    try {
      final actualTestsPath =
          testsPath.startsWith('/') ? testsPath.substring(1) : testsPath;

      final version = test.version + 1;
      final testData = test
          .copyWith(
            steps: test.steps
                .map(
                  (step) => step.copyWith(
                    image: Uint8List.fromList(<int>[]),
                  ),
                )
                .toList(),
            timestamp: DateTime.now(),
            version: version,
          )
          .toJson();

      final testSuitePrefix =
          test.suiteName?.isNotEmpty == true ? '${test.suiteName}/' : '';

      final path = '$actualTestsPath/$testSuitePrefix${test.name}.json';

      final encoder = const JsonEncoder.withIndent('  ');
      await uploadFile(
        data: utf8.encode(encoder.convert(testData)),
        message: 'Updating test',
        path: path,
      );
      result = true;
    } catch (e, stack) {
      _logger.severe('Error writing test', e, stack);
    }
    return result;
  }

  /// Uploads a file to GitHub
  Future<void> uploadFile({
    required List<int> data,
    required String message,
    required String path,
  }) async {
    await _loadData();

    String? sha;
    for (var entry in _tree.entries!) {
      if (path == entry.path) {
        sha = entry.sha;
        break;
      }
    }

    final url = '/repos/${slug.fullName}/contents/${_encodeGitPath(path)}';
    final body = json.encode({
      'branch': branch,
      'committer': _committer,
      'content': base64.encode(data),
      'message': message,
      if (sha != null) 'sha': sha
    });
    final response = await _github.putJSON(
      url,
      body: body,
      headers: {
        'content-type': 'application/json',
      },
    );

    if (response?['commit']?['sha'] == null) {
      throw Exception('[GITHUB UPLOAD]: Error -- [$response]');
    } else {
      _dirty = true;
      _logger.finest(
        '[GITHUB UPLOAD]: [$path] -- [${response?['commit']?['sha']}]',
      );
    }
  }

  Future<void> _loadData() async {
    if (_dirty) {
      _currentGoldenTestImages = null;
      _tree = await _github.git.getTree(slug, branch, recursive: true);
      _dirty = false;
    }
  }

  /// Encodes a path to be suitable to send to GitHub's for the content API.
  String _encodeGitPath(String path) {
    final builder = StringBuffer();

    for (var i = 0; i < path.length; i++) {
      var ch = path.substring(i, i + 1);

      switch (ch) {
        case ' ':
          ch = '%20';
          break;
        case '!':
          ch = '%21';
          break;
        case ':':
          ch = '%3A';
          break;
        case ';':
          ch = '%3B';
          break;
        case '#':
          ch = '%23';
          break;
        case '&':
          ch = '%26';
          break;
        case '?':
          ch = '%3F';
          break;
      }

      builder.write(ch);
    }

    return builder.toString();
  }
}
